pkg/solana/logpoller/loader.go

```diff
@@ -1,477 +1,163 @@
-}
-
-func (c *EncodedLogCollector) start(_ context.Context) error {
-	c.engine.Go(c.runSlotPolling)
-	c.engine.Go(c.runSlotProcessing)
-	c.engine.Go(c.runBlockProcessing)
-	c.engine.Go(c.runJobProcessing)
-
-	return nil
-}
-
-func (c *EncodedLogCollector) close() error {
-	return nil
-}
-
-func (c *EncodedLogCollector) runSlotPolling(ctx context.Context) {
-	for {
-		timer := time.NewTimer(DefaultNextSlotPollingInterval)
-
+	for _, job := range slotsForAddressJobs {
 		select {
 		case <-ctx.Done():
-			timer.Stop()
-
-			return
-		case <-timer.C:
-			ctxB, cancel := context.WithTimeout(ctx, c.rpcTimeLimit)
-
-			// not to be run as a job, but as a blocking call
-			result, err := c.client.GetLatestBlockhash(ctxB, rpc.CommitmentFinalized)
-			if err != nil {
-				c.lggr.Error("failed to get latest blockhash", "err", err)
-				cancel()
-
-				continue
-			}
-
-			cancel()
-
-			// if the slot is not higher than the highest slot, skip it
-			if c.lastSentSlot.Load() >= result.Context.Slot {
-				continue
-			}
-
-			c.lastSentSlot.Store(result.Context.Slot)
-
-			select {
-			case c.chSlot <- result.Context.Slot:
-			default:
-			}
+			return nil, ctx.Err()
+		case <-job.Done():
 		}
-
-		timer.Stop()
 	}
-}
 
-func (c *EncodedLogCollector) runSlotProcessing(ctx context.Context) {
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		case slot := <-c.chSlot:
-			if c.highestSlot.Load() >= slot {
-				continue
-			}
-
-			from := c.highestSlot.Load() + 1
-			if c.highestSlot.Load() == 0 {
-				from = slot
-			}
-
-			c.highestSlot.Store(slot)
-
-			// load blocks in slot range
-			c.loadRange(ctx, from, slot)
-		}
+	// it should be safe to access slotsToFetch without lock as all the jobs signalled that they are done
+	result := make([]uint64, 0, len(slotsToFetch))
+	for slot := range slotsToFetch {
+		result = append(result, slot)
 	}
-}
 
-func (c *EncodedLogCollector) runBlockProcessing(ctx context.Context) {
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		case slot := <-c.chBlock:
-			if err := c.workers.Do(ctx, &getTransactionsFromBlockJob{
-				slotNumber: slot,
-				client:     c.client,
-				parser:     c.ordered,
-				chJobs:     c.chJobs,
-			}); err != nil {
-				c.lggr.Errorf("failed to add job to queue: %s", err)
-			}
-		}
-	}
+	sort.Slice(result, func(i, j int) bool { return result[i] < result[j] })
+	return result, nil
 }
 
-func (c *EncodedLogCollector) runJobProcessing(ctx context.Context) {
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		case job := <-c.chJobs:
-			if err := c.workers.Do(ctx, job); err != nil {
-				c.lggr.Errorf("failed to add job to queue: %s", err)
-			}
+func (c *EncodedLogCollector) scheduleBlocksFetching(ctx context.Context, slots []uint64) (<-chan Block, error) {
+	blocks := make(chan Block)
+	getBlocksJobs := make([]*getBlockJob, len(slots))
+	for i, slot := range slots {
+		getBlocksJobs[i] = newGetBlockJob(c.client, blocks, c.lggr, slot)
+		err := c.workers.Do(ctx, getBlocksJobs[i])
```

Comment:
I'd recommend renaming this to getBlockJobs instead of getBlocksJobs since it's only fetching a single block with GetBlockWithOpts not calling GetBlocks